playground/backend/internal/code_processing/code_processing.go

```diff
@@ -0,0 +1,218 @@
+package code_processing
+
+import (
+	pb "beam.apache.org/playground/backend/internal/api/v1"
+	"beam.apache.org/playground/backend/internal/cache"
+	"beam.apache.org/playground/backend/internal/environment"
+	"beam.apache.org/playground/backend/internal/executors"
+	"beam.apache.org/playground/backend/internal/fs_tool"
+	"beam.apache.org/playground/backend/internal/logger"
+	"beam.apache.org/playground/backend/internal/setuper"
+	"context"
+	"github.com/google/uuid"
+)
+
+// Process validates, compiles and runs code by pipelineId.
+// During each operation updates status of execution and saves it into cache:
+// - In case of processing works more that timeout duration saves playground.Status_STATUS_RUN_TIMEOUT as cache.Status into cache.
+// - In case of code processing has been canceled saves playground.Status_STATUS_CANCELED as cache.Status into cache.
+// - In case of validation step is failed saves playground.Status_STATUS_VALIDATION_ERROR as cache.Status into cache.
+// - In case of compile step is failed saves playground.Status_STATUS_COMPILE_ERROR as cache.Status and compile logs as cache.CompileOutput into cache.
+// - In case of compile step is completed with no errors saves compile output as cache.CompileOutput into cache.
+// - In case of run step is failed saves playground.Status_STATUS_RUN_ERROR as cache.Status and run logs as cache.RunError into cache.
+// - In case of run step is completed with no errors saves playground.Status_STATUS_FINISHED as cache.Status and run output as cache.RunOutput into cache.
+// At the end of this method deletes all created folders.
+func Process(ctx context.Context, cacheService cache.Cache, lc *fs_tool.LifeCycle, compileBuilder *executors.CompileBuilder, pipelineId uuid.UUID, env *environment.Environment, sdk pb.Sdk) {
+	ctxWithTimeout, cancelByTimeoutFunc := context.WithTimeout(ctx, env.ApplicationEnvs.PipelineExecuteTimeout())
+	defer func(lc *fs_tool.LifeCycle) {
```

Comment:
If I will use `cleanUp()` from `RunCode()` (after `go utils.Process()` I guess), then I will delete all prepared for processing files/folders, because the `RunCode()` could be finished before `Process` starts to process code since it works like a goroutine.

Reply:
I think you could use the done channel for this purpose?